evm/contract.go

```diff
@@ -0,0 +1,418 @@
+// Copyright (C) 2022, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+package evm
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"reflect"
+	"strings"
+
+	"github.com/ava-labs/avalanche-tooling-sdk-go/utils"
+	"github.com/ava-labs/subnet-evm/accounts/abi/bind"
+	"github.com/ava-labs/subnet-evm/core/types"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+var ErrFailedReceiptStatus = fmt.Errorf("failed receipt status")
+
+func removeSurroundingParenthesis(s string) (string, error) {
+	s = strings.TrimSpace(s)
+	if len(s) > 0 {
+		if string(s[0]) != "(" || string(s[len(s)-1]) != ")" {
+			return "", fmt.Errorf("expected esp %q to be surrounded by parenthesis", s)
+		}
+		s = s[1 : len(s)-1]
+	}
+	return s, nil
+}
+
+func removeSurroundingBrackets(s string) (string, error) {
+	s = strings.TrimSpace(s)
+	if len(s) > 0 {
+		if string(s[0]) != "[" || string(s[len(s)-1]) != "]" {
+			return "", fmt.Errorf("expected esp %q to be surrounded by parenthesis", s)
+		}
+		s = s[1 : len(s)-1]
+	}
+	return s, nil
+}
+
+func getWords(s string) []string {
+	words := []string{}
+	word := ""
+	parenthesisCount := 0
+	insideBrackets := false
+	for _, rune := range s {
+		c := string(rune)
+		if parenthesisCount > 0 {
+			word += c
+			if c == "(" {
+				parenthesisCount++
+			}
+			if c == ")" {
+				parenthesisCount--
+				if parenthesisCount == 0 {
+					words = append(words, word)
+					word = ""
+				}
+			}
+			continue
+		}
+		if insideBrackets {
+			word += c
+			if c == "]" {
+				words = append(words, word)
+				word = ""
+				insideBrackets = false
+			}
+			continue
+		}
+		if c == " " || c == "," || c == "(" || c == "[" {
+			if word != "" {
+				words = append(words, word)
+				word = ""
+			}
+		}
+		if c == " " || c == "," {
+			continue
+		}
+		if c == "(" {
+			parenthesisCount++
+		}
+		if c == "[" {
+			insideBrackets = true
+		}
+		word += c
+	}
+	if word != "" {
+		words = append(words, word)
+	}
+	return words
+}
+
+func getMap(
```

Comment:
this one is too hard for me to read.  Unittest will be helpful to understand what's going on there